### ISSUE TITLE
ARM64 build support

### ISSUE THREAD

**Issue by lrettig** [spacemeshos/go-spacemesh#2643]
It's not currently possible to build go-spacemesh successfully on arm64 systems including Mac M1. When attempting to build targeting arm64, I get the error in #2642. When attempting to build targeting amd64, I get another set of errors:

```
GOBIN= GOARCH=amd64 go run . -version
# github.com/spacemeshos/post/gpu
../../go/pkg/mod/github.com/spacemeshos/post@v0.0.0-20210726170806-5cf91a7ae467/gpu/api.go:13:13: undefined: ComputeAPIClass
../../go/pkg/mod/github.com/spacemeshos/post@v0.0.0-20210726170806-5cf91a7ae467/gpu/api.go:22:14: undefined: cGetProviders
../../go/pkg/mod/github.com/spacemeshos/post@v0.0.0-20210726170806-5cf91a7ae467/gpu/api.go:87:14: undefined: cStopCleared
../../go/pkg/mod/github.com/spacemeshos/post@v0.0.0-20210726170806-5cf91a7ae467/gpu/api.go:100:47: undefined: cScryptPositions
../../go/pkg/mod/github.com/spacemeshos/post@v0.0.0-20210726170806-5cf91a7ae467/gpu/api.go:126:13: undefined: StopResult
../../go/pkg/mod/github.com/spacemeshos/post@v0.0.0-20210726170806-5cf91a7ae467/gpu/api.go:127:9: undefined: cStop
```

What do we need to do to support building on arm64, ideally natively but possibly in emulation mode as a stopgap?

CC @avive @noamnelke

**Comment by avive** [comment]
I spoke about this with @moshababo  - if there are plans to add support for Linux/ARM64, macOS/ARM64 for go-spacemesh (and maybe even Windows11/ARM64) then we can invest and have gpu-post build libs for these platforms. Note that Github doesn't provide yet macOS/ARM64 and Windows11/ARM64 runners for GHA/CI. From a product perspective, the node support is nice for running on a low-cost computer with less ongoing energy costs (this makes the baseline user costs even lower as its main costs are power), but obviously there lots of optimizations to the node that need to be done to support this kind of runtime environment. e.g. low-power ARM PCs.

**Comment by noamnelke** [comment]
We must decide what platforms we want to support. Setting up runners for GHA is a one-time setup + ongoing cost kind of thing. I don't think it's a big challenge and once it's solved - it's solved for building the C binaries, building the Go binaries and running tests.

Now we should decide :)

**Comment by avive** [comment]
From a product perspective we should focus on getting v0.3 working well on adm64 PCs and macOS/amd64/M1 only. Other platforms are niche in the market. Once Spacemesh works well on these platforms we should consider optimizing and targeting node to low-cost ARM PCs (Linux and Windows OS) as it reduces operation costs for users. For macOS on M1/arm64, Spacemesh/amd64 works well and gpu-post also works so there's no need for a special build for M1/arm64 yet.

So I think we should revisit it once node v0.3 is stable for am64 windows/linux/macOS and prioritize at that time considering with other tasks at this time pre-genesis and arm support is def not a v0.2 feature.

**Comment by noamnelke** [comment]
I understand about Pi support, but not entirely about the M1. What is "Spacemesh/amd64"? If we build on a non-M1 mac doesn't it get built for the intel X86 architecture? Or am I completely not understanding how this works?

**Comment by lrettig** [comment]
Possibly related: https://pkg.go.dev/cmd/cgo#hdr-Using_cgo_with_the_go_command

> The cgo tool is enabled by default for native builds on systems where it is expected to work. It is disabled by default when cross-compiling. You can control this by setting the CGO_ENABLED environment variable when running the go tool: set it to 1 to enable the use of cgo, and to 0 to disable it. The go tool will set the build constraint "cgo" if cgo is enabled. The special import "C" implies the "cgo" build constraint, as though the file also said "// +build cgo". Therefore, if cgo is disabled, files that import "C" will not be built by the go tool. (For more about build constraints see https://golang.org/pkg/go/build/#hdr-Build_Constraints).

> When cross-compiling, you must specify a C cross-compiler for cgo to use. You can do this by setting the generic CC_FOR_TARGET or the more specific CC_FOR_${GOOS}_${GOARCH} (for example, CC_FOR_linux_arm) environment variable when building the toolchain using make.bash, or you can set the CC environment variable any time you run the go tool.

**Comment by lrettig** [comment]
> If we build on a non-M1 mac doesn't it get built for the intel X86 architecture?

Correct - not sure what you're confused about. We obviously already support building on amd64 mac. Aviv claims that M1 (ARM) macs can build and run the amd64 builds in compatibility mode (Rosetta), but I failed to do so on mine - see above and #2642.

**Comment by lrettig** [comment]
I was eventually able to get this working. I documented the steps here: https://gist.github.com/lrettig/935ca5c166e083fa92d1b64f0021c8c0

tldr: building and running is very easy (you can just use `GOARCH`), but getting tests working is more complicated.
